File: packages/cloud-adapters/database/baremetal.test.ts
/**
 * @jest-environment node
 *
 * Runtime contract test for the bare-metal CNPG DatabaseAdapter.
 * Hermetic via stubbed fetch — verifies wiring (URL, verb, body, auth)
 * for provision/describe/delete. The Go contract test
 * (//packages/cloud-adapters/database:database_test) does the deeper
 * round-trip via an in-process K8s API fake.
 */
import { describe, beforeEach, it, expect } from "@jest/globals";

import { newBaremetalAdapter } from "./baremetal";

type FetchCall = { url: string; init: { method?: string; body?: string; headers?: Record<string, string> } };

interface FakeFetch {
  (url: string, init?: unknown): Promise<unknown>;
  calls: FetchCall[];
}

const newFakeFetch = (responder: () => { ok?: boolean; status?: number; _body?: unknown }): FakeFetch => {
  const calls: FetchCall[] = [];
  const fn = ((url: string, init: { method?: string; body?: string; headers?: Record<string, string> } = {}) => {
    calls.push({ url, init });
    const r = responder();
    return Promise.resolve({
      ok: r.ok ?? true,
      status: r.status ?? 200,
      json: () => Promise.resolve(r._body ?? {}),
      text: () =>
        Promise.resolve(typeof r._body === "string" ? r._body : JSON.stringify(r._body ?? {})),
    });
  }) as FakeFetch;
  fn.calls = calls;
  return fn;
};

const baseCfg = {
  apiServer: "https://kubernetes.test:6443",
  token: "test-token",
  namespace: "tenant-acme",
};

const clusterCR = (name: string) => ({
  apiVersion: "postgresql.cnpg.io/v1",
  kind: "Cluster",
  metadata: { name, namespace: "tenant-acme" },
  spec: { instances: 1 },
});

describe("cloud-adapters/database/baremetal — runtime wiring contract", () => {
  let f: FakeFetch;

  beforeEach(() => {
    f = newFakeFetch(() => ({ _body: clusterCR("db1") }));
  });

  it("provisionInstance POSTs a CNPG Cluster CR to /apis/postgresql.cnpg.io/v1/namespaces/<ns>/clusters", async () => {
    const adapter = newBaremetalAdapter({ ...baseCfg, fetch: f as unknown as typeof fetch });
    const h = await adapter.provisionInstance("tenant-acme-db", "us-central1");

    expect(f.calls).toHaveLength(1);
    expect(f.calls[0]!.url).toBe(
      "https://kubernetes.test:6443/apis/postgresql.cnpg.io/v1/namespaces/tenant-acme/clusters",
    );
    expect(f.calls[0]!.init.method).toBe("POST");

    const cr = JSON.parse(f.calls[0]!.init.body as string);
    expect(cr.apiVersion).toBe("postgresql.cnpg.io/v1");
    expect(cr.kind).toBe("Cluster");
    expect(cr.metadata).toEqual({ name: "tenant-acme-db", namespace: "tenant-acme" });
    expect(cr.spec.instances).toBe(1);
    expect(cr.spec.bootstrap.initdb.database).toBe("app");

    expect(h.id).toBe("tenant-acme/tenant-acme-db");
    expect(h.dsn).toBe(
      "postgres://app@tenant-acme-db-rw.tenant-acme.svc.cluster.local:5432/app?sslmode=require",
    );
  });

  it("authorization header carries the bearer token", async () => {
    const adapter = newBaremetalAdapter({ ...baseCfg, fetch: f as unknown as typeof fetch });
    await adapter.provisionInstance("db1", "");
    expect(f.calls[0]!.init.headers!["Authorization"]).toBe("Bearer test-token");
    expect(f.calls[0]!.init.headers!["Content-Type"]).toBe("application/json");
  });

  it("describeInstance GETs /clusters/<name>", async () => {
    const adapter = newBaremetalAdapter({ ...baseCfg, fetch: f as unknown as typeof fetch });
    const h = await adapter.describeInstance("db1");
    expect(f.calls[0]!.url).toBe(
      "https://kubernetes.test:6443/apis/postgresql.cnpg.io/v1/namespaces/tenant-acme/clusters/db1",
    );
    expect(f.calls[0]!.init.method).toBe("GET");
    expect(h.dsn).toBe(
      "postgres://app@db1-rw.tenant-acme.svc.cluster.local:5432/app?sslmode=require",
    );
  });

  it("deleteInstance DELETEs /clusters/<name>", async () => {
    const adapter = newBaremetalAdapter({ ...baseCfg, fetch: f as unknown as typeof fetch });
    await adapter.deleteInstance("db1");
    expect(f.calls[0]!.init.method).toBe("DELETE");
    expect(f.calls[0]!.url).toBe(
      "https://kubernetes.test:6443/apis/postgresql.cnpg.io/v1/namespaces/tenant-acme/clusters/db1",
    );
  });

  it("non-2xx responses surface as errors", async () => {
    f = newFakeFetch(() => ({ ok: false, status: 404, _body: '{"reason":"NotFound"}' }));
    const adapter = newBaremetalAdapter({ ...baseCfg, fetch: f as unknown as typeof fetch });
    await expect(adapter.describeInstance("missing")).rejects.toThrow(/404/);
  });

  it("custom database name flows into both the CR and the DSN", async () => {
    const adapter = newBaremetalAdapter({
      ...baseCfg,
      database: "analytics",
      fetch: f as unknown as typeof fetch,
    });
    const h = await adapter.provisionInstance("db1", "");
    const cr = JSON.parse(f.calls[0]!.init.body as string);
    expect(cr.spec.bootstrap.initdb.database).toBe("analytics");
    expect(h.dsn).toContain("/analytics?");
  });
});

File: packages/cloud-adapters/database/baremetal_test.go
package database

import (
	"context"
	"encoding/json"
	"io"
	"net/http"
	"net/http/httptest"
	"strings"
	"sync"
	"testing"
)

// fakeK8s is a hermetic in-process stand-in for the Kubernetes API
// server. It implements only the verbs the CNPG adapter exercises
// (POST/GET/DELETE on /apis/postgresql.cnpg.io/v1/namespaces/<ns>/
// clusters[/name]) and keeps applied CRs in memory.
type fakeK8s struct {
	mu        sync.Mutex
	clusters  map[string]map[string]any // ns/name → cluster CR
	requests  []requestLog
}

type requestLog struct {
	method, path, body string
}

func newFakeK8s() *fakeK8s {
	return &fakeK8s{clusters: map[string]map[string]any{}}
}

func (f *fakeK8s) ServeHTTP(w http.ResponseWriter, r *http.Request) {
	f.mu.Lock()
	defer f.mu.Unlock()

	body := ""
	if r.Body != nil {
		buf, _ := io.ReadAll(r.Body)
		body = string(buf)
	}
	f.requests = append(f.requests, requestLog{r.Method, r.URL.Path, body})

	prefix := "/apis/postgresql.cnpg.io/v1/namespaces/"
	if !strings.HasPrefix(r.URL.Path, prefix) {
		http.Error(w, "no route", http.StatusNotFound)
		return
	}
	rest := strings.TrimPrefix(r.URL.Path, prefix)
	parts := strings.Split(rest, "/")
	// expected: <ns>/clusters[/name]
	if len(parts) < 2 || parts[1] != "clusters" {
		http.Error(w, "no route", http.StatusNotFound)
		return
	}
	ns := parts[0]

	switch {
	case r.Method == http.MethodPost && len(parts) == 2:
		// POST .../clusters — create
		var cr map[string]any
		if err := json.NewDecoder(strings.NewReader(body)).Decode(&cr); err != nil {
			http.Error(w, err.Error(), http.StatusBadRequest)
			return
		}
		meta, _ := cr["metadata"].(map[string]any)
		name, _ := meta["name"].(string)
		f.clusters[ns+"/"+name] = cr
		writeJSON(w, cr)

	case r.Method == http.MethodGet && len(parts) == 3:
		// GET .../clusters/<name>
		cr, ok := f.clusters[ns+"/"+parts[2]]
		if !ok {
			http.Error(w, `{"reason":"NotFound"}`, http.StatusNotFound)
			return
		}
		writeJSON(w, cr)

	case r.Method == http.MethodDelete && len(parts) == 3:
		// DELETE .../clusters/<name>
		key := ns + "/" + parts[2]
		if _, ok := f.clusters[key]; !ok {
			http.Error(w, `{"reason":"NotFound"}`, http.StatusNotFound)
			return
		}
		delete(f.clusters, key)
		w.WriteHeader(http.StatusOK)

	default:
		http.Error(w, "unhandled: "+r.Method+" "+rest, http.StatusNotImplemented)
	}
}

func writeJSON(w http.ResponseWriter, v any) {
	w.Header().Set("Content-Type", "application/json")
	_ = json.NewEncoder(w).Encode(v)
}

func newFakeCluster(t *testing.T) (*BaremetalAdapter, *fakeK8s, func()) {
	t.Helper()
	fk := newFakeK8s()
	server := httptest.NewServer(fk)
	adapter := NewBaremetalAdapter(server.Client(), server.URL, "fake-token", "tenant-acme", "app")
	return adapter, fk, server.Close
}

func TestBaremetalAdapter_ProvisionDescribeDelete(t *testing.T) {
	adapter, fk, cleanup := newFakeCluster(t)
	defer cleanup()
	ctx := context.Background()

	h, err := adapter.ProvisionInstance(ctx, "tenant-acme-db", "us-central1")
	if err != nil {
		t.Fatalf("ProvisionInstance: %v", err)
	}
	if h.ID != "tenant-acme/tenant-acme-db" {
		t.Errorf("Handle.ID: got %q want tenant-acme/tenant-acme-db", h.ID)
	}
	want := "postgres://app@tenant-acme-db-rw.tenant-acme.svc.cluster.local:5432/app?sslmode=require"
	if h.DSN != want {
		t.Errorf("Handle.DSN: got %q want %q", h.DSN, want)
	}

	// Verify the CR shape the fake K8s API received.
	if len(fk.requests) != 1 {
		t.Fatalf("requests: got %d, want 1", len(fk.requests))
	}
	req := fk.requests[0]
	if req.method != "POST" || !strings.Contains(req.path, "/namespaces/tenant-acme/clusters") {
		t.Errorf("provision request: got %s %s", req.method, req.path)
	}
	var cr map[string]any
	if err := json.NewDecoder(strings.NewReader(req.body)).Decode(&cr); err != nil {
		t.Fatalf("decode CR: %v", err)
	}
	if cr["apiVersion"] != "postgresql.cnpg.io/v1" || cr["kind"] != "Cluster" {
		t.Errorf("CR apiVersion/kind: got %v/%v", cr["apiVersion"], cr["kind"])
	}

	// Describe round-trips through GET.
	h2, err := adapter.DescribeInstance(ctx, "tenant-acme-db")
	if err != nil {
		t.Fatalf("DescribeInstance: %v", err)
	}
	if h2.DSN != want {
		t.Errorf("DescribeInstance DSN: got %q want %q", h2.DSN, want)
	}

	// Delete reaps; subsequent describe surfaces the error path.
	if err := adapter.DeleteInstance(ctx, "tenant-acme-db"); err != nil {
		t.Fatalf("DeleteInstance: %v", err)
	}
	if _, err := adapter.DescribeInstance(ctx, "tenant-acme-db"); err == nil {
		t.Errorf("DescribeInstance after delete: expected error, got nil")
	}
}

func TestBaremetalAdapter_BearerTokenHeader(t *testing.T) {
	var seenAuth string
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		seenAuth = r.Header.Get("Authorization")
		w.Header().Set("Content-Type", "application/json")
		w.Write([]byte(`{"apiVersion":"postgresql.cnpg.io/v1","kind":"Cluster","metadata":{"name":"x","namespace":"y"},"spec":{}}`))
	}))
	defer server.Close()

	adapter := NewBaremetalAdapter(server.Client(), server.URL, "my-token", "ns", "")
	_, err := adapter.ProvisionInstance(context.Background(), "x", "")
	if err != nil {
		t.Fatalf("ProvisionInstance: %v", err)
	}
	if seenAuth != "Bearer my-token" {
		t.Errorf("Authorization header: got %q, want %q", seenAuth, "Bearer my-token")
	}
}
